lib/events.ex

```diff
@@ -1,3 +1,104 @@
 defmodule Events do
-  @moduledoc false
+  @moduledoc """
+  Events allows to attach functions to be called when events are emitted
+
+  Note that all subscribed functions are called in the process which emits the event.
+  """
+
+  require Logger
+
+  @callback_mod Application.fetch_env!(:events, :impl)
+
+  @type handler_id :: term()
+  @type event_name :: [atom()]
+  @type event_value :: number()
+  @type event_prefix :: [atom()]
+
+  @doc """
+  Attaches a handler to the event prefix.
+
+  `handler_id` must be unique, if another handler with the same ID already exists the
+  `:error` atom is be returned.
+
+  When events with `event_prefix` are emitted, function `function` in module `module` will be
+  called with three arguments:
+  * the exact event name (see `execute/2`)
+  * the event value (see `execute/2`)
+  * the handler configuration, `config`, or `nil` if one wasn't provided
+
+  If the function fails (raises, exits or throws) then the handler is removed.
+  """
+  @spec attach(handler_id, event_prefix, module, function :: atom) ::
+          :ok | {:error, :already_exists}
+  @spec attach(handler_id, event_prefix, module, function :: atom, config :: term) ::
+          :ok | {:error, :already_exists}
+  def attach(handler_id, event_prefix, module, function, config \\ nil) do
+    assert_event_name_or_prefix(event_prefix)
+    @callback_mod.attach(handler_id, event_prefix, module, function, config)
+  end
+
+  @doc """
+  Removes existing handler.
+
+  If the handler with given ID doesn't exist, `:error` is returned.
+  """
+  @spec detach(handler_id) :: :ok | {:error, :doesnt_exist}
+  defdelegate detach(handler_id), to: @callback_mod
+
+  @doc """
+  Emits an event, executing handlers attached to all of its prefixes.
+
+  Note that you should not rely on the order in which those functions are called.
+  """
+  @spec execute(event_name, event_value) :: :ok
+  def execute(event_name, value) when is_number(value) do
+    assert_event_name_or_prefix(event_name)
+
+    handlers = @callback_mod.list_handlers_by_name(event_name)
```

Comment:
I thought the dispatching was meant to be done by prefix? If the dispatch is by name, is there a reason to use prefixes?  :)

Reply:
`list_handlers_by_name` means "return all handlers which should be invoked for given event name". `list_handlers_by_prefix` means "return all handlers which are attached to the prefix equal or greater than the given prefix". Maybe these names do not reflect correctly what the functions do. Let me give you an example:

```elixir
iex> Events.attach("id1", [:alice, :has, :a, :cat], MyHandler, :handle)
iex> Events.attach("id2", [:alice, :has], MyHandler, :handle)
iex> Events.attach("id3", [:alice], MyHandler, :handle)

iex> list_handlers_by_name([:alice, :has, :a, :cat])
["id1", "id2", "id3"] # rest of return values skipped for brevity
iex> list_handlers_by_name([:alice, :has])
["id1", "id2", "id3"]
iex> list_handlers_by_name([:alice])
["id1", "id2", "id3"]

iex> list_handlers_by_prefix([:alice, :has, :a, :cat])
["id1"]
iex> list_handlers_by_prefix([:alice, :has])
["id1", "id2"]
iex> list_handlers_by_prefix([:alice])
["id1", "id2", "id3"]
```

Maybe I'll rephrase: `list_handlers_by_name` allows me to see which handlers should be invoked for the given event. `list_handlers_by_prefix` returns handlers attached to some event "namespace" (I imagine one could use it to view all handlers installed by Phoenix by calling `list_handlers_by_name([:phoenix])`.

Does it make sense? Listing by prefix can be used for discovering installed handlers, but you're right that it's not necessarily needed.

Reply:
Ah, great.

The `by_name` give me the impression of filtering by the "name" (i.e. prefix) given to the handler. Maybe `list_handlers_for_event` and `list_handlers_by_prefix`?

Reply:
@josevalim I've renamed the callback. It's definitely better now :)